### PR TITLE
[CARBONDATA-37]Support different time format input style

### PR DESCRIPTION
**Pull request by lion-x**
# Why raise this PR?
support different time format input style. In some scenarios different time dimensions may use different time format, we should support these requirements.

# Description
In this PR, we support load timestamp data col with diff format in every loads.  

statement:

LOAD DATA LOCAL INPATH 'XXXXX' into table tablename
OPTIONS('dateformat' = 'colname1:timeformat1,colname2:timeformat2')

### CODE REVIEW

**Comment by lion-x** [comment]
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/407/

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java [hunk 1]
```diff
@@ -1171,6 +1171,14 @@ else if(isComplexTypeColumn[j]) {
               DirectDictionaryGenerator directDictionaryGenerator1 =
                   DirectDictionaryKeyGeneratorFactory
                       .getDirectDictionaryGenerator(details.getColumnType());
+              String[] timeformats = meta.timeFormat.split(",");
+              for(String timeformat:timeformats){
```
Comment (lion-x) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java]: ok,thx

File: core/src/main/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java [hunk 2]
```diff
@@ -117,15 +117,24 @@ private TimeStampDirectDictionaryGenerator() {
    * @return dictionary value
    */
   @Override public int generateDirectSurrogateKey(String memberStr) {
-    SimpleDateFormat timeParser = new SimpleDateFormat(CarbonProperties.getInstance()
-        .getProperty(CarbonCommonConstants.CARBON_TIMESTAMP_FORMAT,
-            CarbonCommonConstants.CARBON_TIMESTAMP_DEFAULT_FORMAT));
+    String timeString;
+    String formatString;
+    if (memberStr.contains(CarbonCommonConstants.COLON_SPC_CHARACTER)){
```
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java]: It is not a good implement to append format to input data.
No need to combine data with data format to a new member string.

File: core/src/main/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java [hunk 3]
```diff
@@ -117,15 +117,24 @@ private TimeStampDirectDictionaryGenerator() {
    * @return dictionary value
    */
   @Override public int generateDirectSurrogateKey(String memberStr) {
-    SimpleDateFormat timeParser = new SimpleDateFormat(CarbonProperties.getInstance()
-        .getProperty(CarbonCommonConstants.CARBON_TIMESTAMP_FORMAT,
-            CarbonCommonConstants.CARBON_TIMESTAMP_DEFAULT_FORMAT));
+    String timeString;
```
Comment (lion-x) [inline core/src/main/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java]: ok, revised the code by your comment.

**Comment by lion-x** [comment]
By @QiangCai 's comments, Code has been update. 
CI passed
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/483/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 4]
```diff
@@ -1129,6 +1130,9 @@ case class LoadTable(
       carbonLoadModel.setEscapeChar(escapeChar)
       carbonLoadModel.setQuoteChar(quoteChar)
       carbonLoadModel.setCommentChar(commentchar)
+      carbonLoadModel.setDateFormat(dateFormat)
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: It is necessary to validate input "dateFormat" before dataloading

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/dataload/TestLoadDataWithDiffTimestampFormat.scala [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+      )
+      checkAnswer(
+        sql("SELECT date FROM t3 WHERE ID = 18"),
+        Seq(Row(Timestamp.valueOf("2015-07-25 00:00:00.0")))
+      )
+      checkAnswer(
+        sql("SELECT starttime FROM t3 WHERE ID = 18"),
+        Seq(Row(Timestamp.valueOf("2016-07-25 02:32:02.0")))
+      )
+  }
```
Comment (QiangCai) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/dataload/TestLoadDataWithDiffTimestampFormat.scala]: please add more test cases to test wrong dateformat, not exists columns, empty string

File: core/src/main/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java [hunk 6]
```diff
@@ -117,9 +113,11 @@ private TimeStampDirectDictionaryGenerator() {
...
   @Override public int generateDirectSurrogateKey(String memberStr) {
-    SimpleDateFormat timeParser = new SimpleDateFormat(CarbonProperties.getInstance()
-        .getProperty(CarbonCommonConstants.CARBON_TIMESTAMP_FORMAT,
-            CarbonCommonConstants.CARBON_TIMESTAMP_DEFAULT_FORMAT));
+    SimpleDateFormat timeParser = threadLocal.get();
+    if(timeParser == null){
+      timeParser = new SimpleDateFormat(dateFormat);
+      threadLocal.set(timeParser);
+    }
     timeParser.setLenient(false);
```
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java]: Please extract above codes to a new initial method,  and invoke this method in different thread.
It it not good to run these codes in generateDirectSurrogateKey method.

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenMeta.java [hunk 7]
```diff
@@ -651,6 +654,7 @@ public void setDefault() {
     columnSchemaDetails = "";
     columnsDataTypeString="";
     tableOption = "";
+    dateFormat = CarbonCommonConstants.CARBON_TIMESTAMP_DEFAULT_FORMAT;
```
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenMeta.java]: Here should be empty string

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java [hunk 8]
```diff
@@ -470,6 +474,36 @@ public boolean processRow(StepMetaInterface smi, StepDataInterface sdi) throws K
...
+          String[] dateformats = meta.dateFormat.split(",");
+          for (String dateFormat:dateformats) {
+            String[] dateFormatSplits = dateFormat.split(":", 2);
+            dateformatsHashMap.put(dateFormatSplits[0],dateFormatSplits[1]);
+            // TODO  verify the dateFormatSplits is valid or not
+          }
+        }
+        directDictionaryGenerators =
+                new DirectDictionaryGenerator[meta.getDimensionColumnIds().length];
+        for (int i = 0; i < meta.getDimensionColumnIds().length; i++) {
```
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java]: not good to invoke getDimensionColumnIds many times

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java [hunk 9]
```diff
@@ -470,6 +474,36 @@ public boolean processRow(StepMetaInterface smi, StepDataInterface sdi) throws K
...
+                      DirectDictionaryKeyGeneratorFactory.getDirectDictionaryGenerator(
+                              columnSchemaDetails.getColumnType(),
+                              dateformatsHashMap.get(columnSchemaDetails.getColumnName()));
+            } else {
+              String dateFormat = CarbonProperties.getInstance()
+                      .getProperty(CarbonCommonConstants.CARBON_TIMESTAMP_FORMAT,
+                              CarbonCommonConstants.CARBON_TIMESTAMP_DEFAULT_FORMAT);
+              directDictionaryGenerators[i] =
+                      DirectDictionaryKeyGeneratorFactory.getDirectDictionaryGenerator(
+                              columnSchemaDetails.getColumnType(), dateFormat);
```
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java]: 1. move out CarbonProperties.getInstance().getProperty  from for loop
2. for defaut  dataformat, use  method getDirectDictionaryGenerator(DataType dataType)

File: core/src/main/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java [hunk 10]
```diff
@@ -39,37 +39,32 @@
...
+
   private static final LogService LOGGER =
-      LogServiceFactory.getLogService(TimeStampDirectDictionaryGenerator.class.getName());
+          LogServiceFactory.getLogService(TimeStampDirectDictionaryGenerator.class.getName());
 
-  /**
-   * initialization block for granularityFactor and cutOffTimeStamp
-   */
-  static {
+  public TimeStampDirectDictionaryGenerator(String dateFormat) {
```
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGenerator.java]: please keep default dateformat TimeStampDirectDictionaryGenerator() construct method, If DataLoading command didn't provide dateformat option for some column, we can use none-parameter construct method.

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenMeta.java [hunk 11]
```diff
@@ -111,7 +110,7 @@
   /**
    * timeFormat
```
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenMeta.java]: please correct comment

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenMeta.java [hunk 12]
```diff
@@ -111,7 +110,7 @@
...
    */
-  protected SimpleDateFormat timeFormat;
+  protected String dateFormat;
```
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenMeta.java]: better to complete modification for all method and keep the order.
e.g.  saveRep, readRep...

File: processing/src/test/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGeneratorTest.java [hunk 13]
```diff
@@ -37,7 +37,7 @@
   private int surrogateKey = -1;
 
   @Before public void setUp() throws Exception {
-    TimeStampDirectDictionaryGenerator generator = TimeStampDirectDictionaryGenerator.instance;
+    TimeStampDirectDictionaryGenerator generator = new TimeStampDirectDictionaryGenerator(CarbonCommonConstants.CARBON_TIMESTAMP_DEFAULT_FORMAT);
```
Comment (QiangCai) [inline processing/src/test/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGeneratorTest.java]: Should use carbon property to  create generator, not default value.
please correct all.

Comment (QiangCai) [inline processing/src/test/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGeneratorTest.java]: Should use carbon property to  create generator, not default value.
please correct all.

Comment (lion-x) [inline processing/src/test/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGeneratorTest.java]: This file is a test file, I think the TimeStampDirectDictionaryGenerator should be set 'CarbonCommonConstants.CARBON_TIMESTAMP_DEFAULT_FORMAT' for testing. pls check again.

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenMeta.java (hunk 7 above)
Comment (lion-x) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenMeta.java]: ok

**Comment by lion-x** [comment]
code has been updated.
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/496/

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 14]
```diff
@@ -1143,6 +1141,21 @@ case class LoadTableUsingKettle(
       val allDictionaryPath = options.getOrElse("all_dictionary_path", "")
       val complex_delimiter_level_1 = options.getOrElse("complex_delimiter_level_1", "\\$")
       val complex_delimiter_level_2 = options.getOrElse("complex_delimiter_level_2", "\\:")
+      val timeFormat = options.getOrElse("timeformat", null)
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: "timeFormat" is useless

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 15]
```diff
@@ -1156,6 +1169,9 @@ case class LoadTableUsingKettle(
       carbonLoadModel.setEscapeChar(escapeChar)
       carbonLoadModel.setQuoteChar(quoteChar)
       carbonLoadModel.setCommentChar(commentchar)
+      carbonLoadModel.setDateFormat(dateFormat)
+      carbonLoadModel.setSerializationNullFormat("serialization_null_format" + "," +
+        serializationNullFormat)
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: this code is useless

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 16]
```diff
@@ -1143,6 +1141,21 @@ case class LoadTableUsingKettle(
...
+        dimension => {
+          val columnSchema: ColumnSchema = dimension.asInstanceOf[CarbonDimension].getColumnSchema
+          if (columnSchema.getDataType.name == "TIMESTAMP") {
+            dateDimensionsName += columnSchema.getColumnName
+          }
+        }
+      }
+      if (dateFormat != null) {
+        validateDateFormat(dateFormat, dateDimensionsName)
+      }
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: please move these code into method validateDateFormat

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 17]
```diff
@@ -1244,6 +1260,25 @@ case class LoadTableUsingKettle(
...
 
+  private def validateDateFormat(dateFormat: String, dateDimensionsName: ArrayBuffer[String]):
+  Unit = {
+    if (dateFormat == "") {
+      throw new MalformedCarbonCommandException("Error: Option DateFormat is set an empty string.")
+    } else {
+      var dateFormats: Array[String] = dateFormat.split(",")
+      for (singleDateFormat <- dateFormats) {
+        var dateFormatSplits: Array[String] = singleDateFormat.split(":", 2)
+        if (!dateDimensionsName.contains(dateFormatSplits(0))) {
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: take care case-insensitive

File: processing/src/test/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGeneratorTest.java (hunk 13 above)
Comment (lion-x) [inline processing/src/test/java/org/apache/carbondata/core/keygenerator/directdictionary/timestamp/TimeStampDirectDictionaryGeneratorTest.java]: ok

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenMeta.java (hunk 11 above)
Comment (lion-x) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenMeta.java]: ok

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/dataload/TestLoadDataWithDiffTimestampFormat.scala [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+           """)
+      assert(false)
+    } catch {
+      case ex: MalformedCarbonCommandException =>
+        assertResult(ex.getMessage)("Error: Wrong Column Name fasfdas is provided in Option DateFormat.")
+      case _ => assert(false)
+    }
+  }
+
+  override def afterAll {
```
Comment (QiangCai) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/dataload/TestLoadDataWithDiffTimestampFormat.scala]: dateformat='startTime'
dateformat='startTime : '
dateformat=' startTime : yyyy-MM-dd HH:mm:ss '
dateformat='startTime:ydm1'
dateformat=':yyyy/MM/dd'

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java [hunk 19]
```diff
@@ -1176,11 +1215,8 @@ else if(isComplexTypeColumn[j]) {
             }
             ColumnSchemaDetails details = columnSchemaDetailsWrapper.get(dimensionColumnIds[m]);
             if (details.isDirectDictionary()) {
-              DirectDictionaryGenerator directDictionaryGenerator1 =
-                  DirectDictionaryKeyGeneratorFactory
-                      .getDirectDictionaryGenerator(details.getColumnType());
               surrogateKeyForHrrchy[0] =
-                  directDictionaryGenerator1.generateDirectSurrogateKey(tuple);
+              directDictionaryGenerators[m].generateDirectSurrogateKey(tuple);
```
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java]: take care code style

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java [hunk 20]
```diff
@@ -343,7 +345,8 @@ public boolean processRow(StepMetaInterface smi, StepDataInterface sdi) throws K
           }
 
           data.setGenerator(
-              KeyGeneratorFactory.getKeyGenerator(getUpdatedLens(meta.dimLens, meta.dimPresent)));
+              KeyGeneratorFactory.getKeyGenerator(
+                  getUpdatedLens(meta.dimLens, meta.dimPresent)));
```
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java]: keep code style

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 21]
```diff
@@ -1244,6 +1260,25 @@ case class LoadTableUsingKettle(
     Seq.empty
   }
 
+  private def validateDateFormat(dateFormat: String, dateDimensionsName: ArrayBuffer[String]):
+  Unit = {
+    if (dateFormat == "") {
+      throw new MalformedCarbonCommandException("Error: Option DateFormat is set an empty string.")
+    } else {
+      var dateFormats: Array[String] = dateFormat.split(",")
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: CarbonCommonConstant.COMMA

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java (hunk 19 above)
Comment (lion-x) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java]: ok

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java (hunk 20 above)
Comment (lion-x) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java]: ok

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala (hunk 17 above)
Comment (lion-x) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: ok

File: integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/dataload/TestLoadDataWithDiffTimestampFormat.scala (hunk 5 above)
Comment (lion-x) [inline integration/spark/src/test/scala/org/apache/carbondata/spark/testsuite/dataload/TestLoadDataWithDiffTimestampFormat.scala]: ok

**Comment by lion-x** [comment]
code has been updated.
CI pass
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/514/
@QiangCai pls,review

**Comment by Jay357089** [comment]
LGTM

File: processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java [hunk 22]
```diff
@@ -470,6 +472,34 @@ public boolean processRow(StepMetaInterface smi, StepDataInterface sdi) throws K
...
+        String[] DimensionColumnIds = meta.getDimensionColumnIds();
+        directDictionaryGenerators =
+            new DirectDictionaryGenerator[DimensionColumnIds.length];
+        for (int i = 0; i < DimensionColumnIds.length; i++) {
+          ColumnSchemaDetails columnSchemaDetails = columnSchemaDetailsWrapper.get(
+              DimensionColumnIds[i]);
+          if (columnSchemaDetails.isDirectDictionary()) {
+            String columnName = columnSchemaDetails.getColumnName();
+            DataType columnType = columnSchemaDetails.getColumnType();
+            if (dateformatsHashMap.containsKey(columnName)) {
```
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/surrogatekeysgenerator/csvbased/CarbonCSVBasedSeqGenStep.java]: better to use "get" method, just look up map once.

File: integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala [hunk 23]
```diff
@@ -1244,6 +1246,29 @@ case class LoadTableUsingKettle(
     Seq.empty
   }
 
+  private  def validateDateFormat(dateFormat: String, table: CarbonTable): Unit = {
+    val dimensions = table.getDimensionByTableName(tableName).asScala
+    if (dateFormat != null) {
+      if (dateFormat.trim == "") {
+        throw new MalformedCarbonCommandException("Error: Option DateFormat is set an empty " +
```
Comment (QiangCai) [inline integration/spark/src/main/scala/org/apache/spark/sql/execution/command/carbonTableSchema.scala]: better to remove "Error: " for all exception message

**Comment by lion-x** [comment]
code has been updated.
CI pass
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/533/

**Comment by QiangCai** [comment]
LGTM
